kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -142,110 +134,111 @@ class KinesisSink private (
-                  val exampleMsg = items.map(_._2.message).find(_.nonEmpty).getOrElse("")
-                  log.error(
-                    s"Failed writing ${items.size} records to SQS queue ${sqs.sqsBufferName}, with error code [$errorCode] and example message [$exampleMsg]."
-                  )
-              }
-              val errorMessage = {
-                if (retriesLeft > 1)
-                  s"Retrying all messages that could not be written to SQS queue ${sqs.sqsBufferName} in $nextBackoff milliseconds..."
-                else
-                  s"Sending ${s.size} messages that could not be written to SQS queue ${sqs.sqsBufferName} to Kinesis stream $streamName in $nextBackoff milliseconds..."
-              }
-              log.error(errorMessage)
-              scheduleWrite(s.map(_._1), target, nextBackoff)(retriesLeft - 1)
-            }
-          case Failure(f) =>
-            log.error("Writing failed with error:", f)
+  def writeBatchToSqsWithRetries(
+    batch: List[Events],
+    sqs: SqsClientAndName,
+    nextBackoff: Long,
+    retriesLeft: Int
+  ): Unit = {
+    log.info(s"Writing ${batch.size} records to SQS buffer ${sqs.sqsBufferName}")
+    writeBatchToSqs(batch, sqs).onComplete {
+      case Success(s) =>
+        sqsHealthy = true
+        log.info(
+          s"Successfully wrote ${batch.size - s.size} out of ${batch.size} records to SQS buffer ${sqs.sqsBufferName}"
+        )
+        if (s.nonEmpty) {
+          s.groupBy(_._2.code).foreach {
+            case (errorCode, items) =>
+              val exampleMsg = items.map(_._2.message).find(_.nonEmpty).getOrElse("")
+              log.error(
+                s"Writing ${items.size} records to SQS buffer ${sqs.sqsBufferName} failed with error code [$errorCode] and example message: $exampleMsg"
+              )
+          }
+          val failedRecords = s.map(_._1)
+          handleSqsError(failedRecords, sqs, nextBackoff, retriesLeft)
+        }
+      case Failure(f) =>
+        log.error(
+          s"Writing ${batch.size} records to SQS buffer ${sqs.sqsBufferName} failed with error: ${f.getMessage()}"
+        )
+        handleSqsError(batch, sqs, nextBackoff, retriesLeft)
+    }
+  }
 
-            if (retriesLeft > 1) {
-              log.error(s"Retrying writing batch to SQS queue ${sqs.sqsBufferName} in $nextBackoff milliseconds...")
-            } else {
-              outage = true
-              log.error(s"Sending batch to Kinesis stream $streamName in $nextBackoff milliseconds...")
-            }
+  def handleKinesisError(failedRecords: List[Events], nextBackoff: Long, retriesLeft: Int): Unit =
+    if (retriesLeft > 0) {
+      log.error(
+        s"$retriesLeft retries left. Retrying to write ${failedRecords.size} records to Kinesis stream $streamName in $nextBackoff milliseconds"
+      )
+      scheduleRetryToKinesis(failedRecords, nextBackoff, retriesLeft - 1)
+    } else {
+      kinesisHealhy = false
```

Comment:
This line surprised me the most.  It is possible that all requests to kinesis were "successful" (returned a success response) but we ran out of retries because we got throttled by kinesis rate limits too many times.

So this changes our definition of unhealthy.  In previous versions, the sink was only marked as unhealthy if kinesis returned error responses.  Whereas now the sink could be marked unhealthy during a traffic spike because of throttling.

I think that might be OK... because we still have the sqs buffer, which sets overall health to healthy.  So I'm just checking this change is deliberate, and you have fully thought through the consequences?

Reply:
> This line surprised me the most. It is possible that all requests to kinesis were "successful" (returned a success response) but we ran out of retries because we got throttled by kinesis rate limits too many times.

Yes it's possible!

> In previous versions, the sink was only marked as unhealthy if kinesis returned error responses. Whereas now the sink could be marked unhealthy during a traffic spike because of throttling.

As I understand [there was already no distinction between throughput exceeded errors and internal errors](https://github.com/snowplow/stream-collector/blob/2.8.2/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala#L187-L204). Am I missing something?

> I think that might be OK... because we still have the sqs buffer, which sets overall health to healthy.

I also think that this is how the behavior should be : if we hit throughput exceeded 3 times, we are in a spike, and we want to write to SQS in this case, if set up.

Reply:
> As I understand [there was already no distinction between throughput exceeded errors and internal errors](https://github.com/snowplow/stream-collector/blob/2.8.2/kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala#L187-L204). Am I missing something?

I don't read it that way, but maybe I'm reading it wrong.  In the old version, if you look for the lines of code which set `outage = true`, I think it never happens if Kinesis always returns a successful result.

But don't worry about it -- this change (if it is a change) only affects collectors with no sqs buffer.  For collectors with a sqs buffer then I'm more confident there is no change compared to the old behaviour.

Reply:
I think that it's a good change for collector without SQS buffer.

The variable is now named `kinesisHealthy` (unlike `outage` before), like `/sink-health`, and if writes are throttled, I think that it's not healthy behavior and we should know about it, because the stream might be under provisioned.

Reply:
That's good reasoning, I'm happy with that.